File: sdk/java-sdk-protobuf/src/main/java/kalix/javasdk/valueentity/ValueEntityProvider.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.javasdk.valueentity;

import kalix.javasdk.Kalix;
import kalix.javasdk.impl.MessageCodec;
import kalix.javasdk.impl.valueentity.ValueEntityRouter;
import com.google.protobuf.Descriptors;

import java.util.Optional;

/**
 * Register a value based entity in {@link Kalix} using a <code>
 * ValueEntityProvider</code>. The concrete <code>ValueEntityProvider</code> is generated for the
 * specific entities defined in Protobuf, for example <code>CustomerEntityProvider</code>.
 */
public interface ValueEntityProvider<S, E extends ValueEntity<S>> {

  ValueEntityOptions options();

  Descriptors.ServiceDescriptor serviceDescriptor();

  String typeId();

  ValueEntityRouter<S, E> newRouter(ValueEntityContext context);

  Descriptors.FileDescriptor[] additionalDescriptors();

  default Optional<MessageCodec> alternativeCodec() {
    return Optional.empty();
  }
}

File: sdk/scala-sdk-protobuf/src/main/scala/kalix/scalasdk/valueentity/ValueEntityProvider.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.scalasdk.valueentity

import kalix.scalasdk.impl.valueentity.ValueEntityRouter
import com.google.protobuf.Descriptors

/**
 * Register a value based entity in {@link kalix.scalasdk.Kalix} using a <code> ValueEntityProvider</code>. The concrete
 * <code>ValueEntityProvider</code> is generated for the specific entities defined in Protobuf, for example
 * <code>CustomerEntityProvider</code>.
 */
trait ValueEntityProvider[S, E <: ValueEntity[S]] {
  def options: ValueEntityOptions

  def serviceDescriptor: Descriptors.ServiceDescriptor

  def typeId: String

  def newRouter(context: ValueEntityContext): ValueEntityRouter[S, E]

  def additionalDescriptors: Seq[Descriptors.FileDescriptor]
}
